feat(windows): add custom Windows title bar to replace the default

SilkePilon opened the pull request:
- Introduce a new WindowsTitleBar component for a native-like title
  bar experience on Windows platform.
- Conditionally render WindowsTitleBar in the main layout when running
  on Windows.
- Modify window creation logic to disable default window decorations
  (title bar) on Windows, enabling the custom title bar to be used.
- Keep default decorations on other platforms to preserve native look.
- Improve app branding by showing version and build type badges in the
  custom title bar.
- Enhance user experience with zoom controls and window management
  integrated into the custom title bar.

This change improves UI consistency and usability for Windows users by
providing a tailored window chrome that matches platform conventions.

- [ ] Linux tests
- [x] MaxOS tests
- [x] Windows tests

SilkePilon committed:
feat(windows): add custom Windows title bar and hide default

- Introduce a new WindowsTitleBar component for a native-like title
  bar experience on Windows platform.
- Conditionally render WindowsTitleBar in the main layout when running
  on Windows.
- Modify window creation logic to disable default window decorations
  (title bar) on Windows, enabling the custom title bar to be used.
- Keep default decorations on other platforms to preserve native look.
- Improve app branding by showing version and build type badges in the
  custom title bar.
- Enhance user experience with zoom controls and window management
  integrated into the custom title bar.

This change improves UI consistency and usability for Windows users by
providing a tailored window chrome that matches platform conventions.

SilkePilon committed:
Merge pull request #2 from SilkePilon/feat-windows-custom-titlebar

Byron committed:
Apply `cargo fmt` and auto-fix node lints

SilkePilon commented:
The current issue is i cant build or preview my changes on my pc atm, as i need to restart my pc every time i make an change as the port becomes locked

SilkePilon committed:
feat: add WindowsTitleBar component with app info and controls

Introduce a new WindowsTitleBar.svelte component for the desktop app that
displays app version, build type badge, and app icon. Implement zoom controls,
theme switching, sidebar toggling, project management actions, and developer
tools access. Detect build type dynamically and update UI accordingly.

This enhances user experience on Windows by providing quick access to common
actions and app status directly in the title bar.

SilkePilon committed:
Merge branch 'master' into feat-add-windows-titlebar-component

SilkePilon committed:
Merge pull request #3 from SilkePilon/feat-add-windows-titlebar-component

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
feat(windows-title-bar): add keyboard shortcuts and reactive project

- Replace static project context with reactive project state updated on route
  changes to reflect the active project dynamically.
- Import and integrate KeyboardShortcutsModal component to display shortcuts UI.
- Register global keyboard shortcuts for common actions (add repo, clone, settings,
  theme switch, zoom, reload, share debug) via ShortcutService.
- Enhance File menu with keyboard shortcut hints and auto-close behavior on selection.
- Remove unused toggleSidebar function and related event emission.
- Improve user experience by enabling quick access to features through shortcuts and
  keeping project context in sync with navigation.

SilkePilon committed:
Merge pull request #4 from SilkePilon/feat-add-windows-titlebar-component

SilkePilon committed:
feat(window): hide default title bar on Windows

Disable window decorations to remove the default title bar on Windows,
enabling a custom title bar implementation. Also, update the maximize
button SVG markup for improved readability and consistency in the
WindowsTitleBar component.

SilkePilon committed:
Merge pull request #5 from SilkePilon/feat-add-windows-titlebar-component

SilkePilon commented:
Fully working on windows 1:1

SilkePilon committed:
feat(windows-title-bar): add editor badge to title bar

Display the default code editor's name as a badge in the Windows title
bar when set in user settings. This improves visibility of the selected
editor and enhances user context. Integrate the badge snippet into the
context menu control for consistent UI behavior.

SilkePilon committed:
Merge pull request #6 from SilkePilon/feat/windows-title-bar-editor-badge

SilkePilon commented:
Added an preview badge to display in what IDE the project gets opened

![image](https://github.com/user-attachments/assets/496f3fb6-0da5-42bf-909e-8e8d9a6d74be)

SilkePilon commented:
The custom titlebar is now optional (will be moved down in the settings menu in the future)

https://github.com/user-attachments/assets/a702f0fd-048c-4b98-adb1-2e54a2321fb7

SilkePilon committed:
Add a new user setting `useCustomTitleBar` to enable or disable the
custom title bar on Windows. Update the Tauri backend to support
setting window decorations dynamically via a new `setDecorations`
method.

Initialize window decorations on app start based on the user setting,
and update decorations immediately when the setting changes. Change
default window creation to start with decorations enabled, allowing
the frontend to control visibility.

This improves user control over the window appearance and aligns the
UI with user preferences for title bar styling on Windows.

SilkePilon committed:
refactor: move Windows title bar toggle to bottom of appearance settings

Relocate the "Use custom title bar" toggle for Windows platform from the
middle to the bottom of the Appearance Settings section to improve UI
layout and consistency. This change does not alter functionality but
enhances the visual structure of the settings page.

fix: correct typo in WindowsTitleBar editor display name lookup

Fix the arrow function formatting in the editor display name lookup to
ensure consistent code style and prevent potential parsing issues.

SilkePilon committed:
Merge branch 'master' into add-custom-titlebar-setting

SilkePilon committed:
Merge pull request #7 from SilkePilon/add-custom-titlebar-setting

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
fix(desktop): apply padding only with custom title bar on Windows

Update app-root padding to apply only when the custom title bar is shown,
preventing unnecessary top padding on other platforms or default title bars.
Add conditional has-custom-titlebar class based on platform and user settings
to control this styling dynamically. This ensures content does not overlap
the title bar only when needed, improving layout consistency.

SilkePilon committed:
Merge pull request #8 from SilkePilon/fix-windows-custom-titlebar-padding

PavelLaptev commented:
@SilkePilon here are colors that I used:

- Header background: `--clr-bg-3`
- text and icons: `--clr-text-1` 

![image](https://github.com/user-attachments/assets/7343d72b-4d07-4518-ad82-3e4325c5c9c4)

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
feat(ui,desktop): add borderRadius to Badge and improve title bar menus

- Add optional borderRadius prop to Badge component for customizable
  corner rounding; default to 20px for consistent styling.
- Update WindowsTitleBar to bind dropdown menu components and handle
  their show events on click, improving menu interaction.
- Increase app icon resolution to 128x128 for nightly, dev, and stable
  builds to enhance visual clarity.
- Replace app icon container div with direct img element and update
  class for better styling control.
- Add data-tauri-drag-region="false" to menu and native window control
  containers to prevent drag conflicts.
- Rename window control button classes to native-control-button for
  clearer semantics and styling consistency.

SilkePilon committed:
Merge pull request #9 from SilkePilon/feat-ui-borderradius-titlebar-menus

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
feat(windows): always use custom title bar and remove toggle

Always show the custom title bar on Windows by default, removing the
user setting and toggle to enable or disable it. Update window
decorations initialization to always hide the default title bar on
Windows. This simplifies the UI and ensures a consistent appearance
across Windows installations.

SilkePilon committed:
style(desktop): adjust WindowsTitleBar menu button spacing

Update dropdown menu styles to remove margin on the wrapper and 
increase horizontal padding on menu buttons for better alignment. 
Set width to fit-content and min-width to auto to improve button 
sizing and layout consistency. These changes enhance the visual 
appearance and usability of the title bar menu.

SilkePilon committed:
refactor: reorganize imports and streamline project state logic

Rearrange import statements in WindowsTitleBar.svelte for better
readability and grouping by functionality. Simplify the reactive effect
that updates the active project on route changes by using promise
chaining instead of an async function. Remove redundant comments and
unused variables to clean up the component state initialization.

These changes improve code clarity andability without altering
component behavior.

SilkePilon committed:
Merge pull request #10 from SilkePilon/windows-custom-titlebar-ui-refactor

SilkePilon committed:
fix(desktop): remove extraneous closing braces in WindowsTitleBar

Remove unnecessary closing braces in the ContextMenuItem component
to a syntax issue and ensure proper rendering of the context menu
items in the Windows title bar. This change improves code clarity and
prevents potential runtime errors.

SilkePilon committed:
Merge pull request #11 from SilkePilon/fix-windows-titlebar-braces

mtsgrd commented on apps/desktop/src/lib/backend/tauri.ts:
```diff
@@ -1,10 +1,37 @@
+	}
+
+	async close() {
+		const window = getCurrentWindow();
```
Since we work with an instance of this class, could we make `window` a private class field? `private window = getCurrentWindow()`

SilkePilon replied:
Made `window` a private class field (`private window = getCurrentWindow()`) and updated all methods to use `this.window` instead of calling `getCurrentWindow()` repeatedly. This improves performance and follows better OOP practices.

mtsgrd commented on apps/desktop/src/lib/backend/tauri.ts:
```diff
@@ -1,10 +1,37 @@
 	checkUpdate = check;
 	currentVersion = getVersion;
+
+	// Window control methods
```
I'm a bit allergic to comments that don't really convey new information. Could we maybe drop this one?

mtsgrd commented on packages/ui/src/lib/Badge.svelte:
```diff
@@ -24,15 +25,21 @@
 	<!-- svelte-ignore a11y_click_events_have_key_events -->
-	<div class="badge {style} {kind} {size}-size" class:reversedDirection {onclick}>
+	<div
+		class="badge {style} {kind} {size}-size"
```
To be consistent with the rest of the code base, could you declare as follows?
```ts
<div
    class="badge"
    class:style
    class:kind
    ...
>
...
</div>
```

SilkePilon replied:
Reverted the `borderRadius` prop addition from Badge.svelte entirely and formatted the component to match the codebase conventions with proper line spacing and indentation.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+	import ContextMenuSection from '@gitbutler/ui/ContextMenuSection.svelte';
+	import DropDownButton from '@gitbutler/ui/DropDownButton.svelte';
+	import KeyboardShortcutsModal from '$components/KeyboardShortcutsModal.svelte';
+	import type { Writable } from 'svelte/store';
```
It might be that you need to build `packages/no-relative-imports` for vscode to automatically sort your imports?

SilkePilon replied:
Built the `packages/no-relative-imports` package which resolved the ESLint/import sorting issues. Imports are now properly merged and sorted in WindowsTitleBar.svelte.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+
+	// Update project when page route changes
+	$effect(() => {
+		$page.url.pathname; // Reactive dependency
+		projectsService
+			.getActiveProject()
+			.then((activeProject) => (project = activeProject))
+			.catch(() => (project = undefined));
```
I'm a bit confused by this, could you help me understand what you are solving for?

SilkePilon replied:
Improved the comment to "Ensures project updates when route changes" which better explains that this creates a reactive dependency on the pathname to trigger project reloading when navigation occurs.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+	});
+
+	// Helper functions
+	const getBadgeText = () => {
+		switch (buildType) {
+			case 'nightly':
+				return `Nightly v${appVersion}`;
+			case 'dev':
+				return 'Development';
+			default:
+				return appVersion || 'Stable';
+		}
+	};
```
For consistency with the rest of the code, could you refactor this into a named function pls?

SilkePilon replied:
Refactored the project update $effect to call a named function `updateProjectFromRoute()` for consistency with the rest of the codebase.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+		}
+	};
+
+	const getBadgeStyle = () => {
+		switch (buildType) {
+			case 'nightly':
+				return 'warning';
+			case 'dev':
+				return 'error';
+			default:
+				return 'neutral';
+		}
+	};
```
Same here.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+	// Zoom functionality
+	const ZOOM_CONFIG = { MIN: 0.375, MAX: 3, DEFAULT: 1, STEP: 0.0625 };
+
+	function updateZoom(newZoom: number) {
+		const zoom = Math.min(Math.max(newZoom, ZOOM_CONFIG.MIN), ZOOM_CONFIG.MAX);
+		document.documentElement.style.fontSize = zoom + 'rem';
+		userSettings.update((s) => ({ ...s, zoom }));
+	}
```
What would it take to reuse existing zoom code here?

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+		zoomIn: () => updateZoom($userSettings.zoom + ZOOM_CONFIG.STEP),
+		zoomOut: () => updateZoom($userSettings.zoom - ZOOM_CONFIG.STEP),
+		resetZoom: () => updateZoom(ZOOM_CONFIG.DEFAULT),
+		openDevTools: () => import.meta.env.DEV && console.log('Opening developer tools...'),
```
👀

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+
+{#snippet editorBadgeSnippet()}
+	{#if editorDisplayName}
+		<Badge style="neutral" size="icon" borderRadius="var(--radius-s)">
```
We should avoid passing css values through props like this, but I'm also wondering if we really want to make the badge radius configurable. @PavelLaptev could you help out here?

SilkePilon replied:
Removed the `borderRadius` prop from Badge.svelte entirely and removed its usage in WindowsTitleBar.svelte. The Badge component now uses its default styling without prop-based customization.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+	</div>
+{/if}
+
+<KeyboardShortcutsModal bind:this={keyboardShortcutsModal} />
```
Is there any way we can avoid having a second instance of this modal? It's currently present in `ChromeSidebar.svelte`.

SilkePilon replied:
Removed the `KeyboardShortcutsModal` import, state, and component usage from WindowsTitleBar.svelte. The keyboard shortcuts action now relies on the global shortcut registration system instead of maintaining a local modal instance, avoiding duplication with the modal in ChromeSidebar.svelte.

mtsgrd commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -0,0 +1,555 @@
+	}
+
+	/* Hide dropdown icons, separators, and vertical lines */
+	.title-bar__menu :global(.dropdown-wrapper .btn .icon),
+	.title-bar__menu :global(.dropdown-wrapper .btn .separator),
+	.title-bar__menu :global(.dropdown-wrapper .btn svg),
+	.title-bar__menu :global(.dropdown-wrapper .btn::after),
+	.title-bar__menu :global(.dropdown-wrapper::after),
+	.title-bar__menu :global(.separator) {
```
Could you look at customizing the dropdown through props rather than overriding css using `:global`?

SilkePilon replied:
Checked the DropDownButton component and confirmed it doesn't support prop-based visual customization for title bar styling. The global CSS overrides are necessary but have been made more targeted and removed the `!important` declarations. The styling is now scoped specifically to `.title-bar__menu` context and follows a cleaner approach.

mtsgrd reviewed (commented):
Awesome stuff, sorry for leaving so many comments(!)

Could you also address the check failures from github actions? Unfortunately eslint will fail silently until you build `packages/no-relative-imports`.

SilkePilon committed:
refactor: improve window handling and code organization

Optimize window handling in tauri.ts by creating a shared window
instance. Extract initializeAppInfo into a separate function and
convert other helper functions to regular function declarations
for better organization. Remove unused imports and variables.

Make window operations more efficient by reusing the window
reference instead of repeatedly getting the current window.
Improve comment clarity for context dependencies in effect hooks.

SilkePilon committed:
Merge pull request #12 from SilkePilon/refactor-improve-window-handling-organization

SilkePilon committed:
refactor(ui): improve badge components and title bar UI

Refactor Badge.svelte to use individual class bindings instead of
string interpolation for better clarity and type safety. Reorder 
Tauri imports for consistency.

Fix warning in WindowsTitleBar.svelte by adding void operator to 
pathname dependency. Upgrade console.log to console.warn for 
developer tools message to increase visibility.

Add missing style and kind attributes to DropDownButtons in the 
title bar menu to ensure consistent styling. Add explicit type="button" 
attributes to window control buttons for better accessibility.

SilkePilon committed:
refactor: improve WindowsTitleBar component zoom handling

Extract updateProjectFromRoute function to improve readability.
Refactor zoom functionality by centralizing logic into dedicated 
functions with defined constants for zoom limits and steps. This
reduces code duplication and improves maintainability by ensuring
consistent zoom behavior across the application.

SilkePilon committed:
Merge pull request #13 from SilkePilon/refactor-badge-and-titlebar-components

SilkePilon committed:
Merge branch 'master' into master

SilkePilon commented:
@mtsgrd Thanks for the review, resolved most of the issues.

SilkePilon commented:
@krlvi — this seems to be a Windows 10-specific issue; it works fine on Windows 11. I’ll look into it, though given that Windows 10 is nearing end-of-life, we might not need to address it unless it affects more users.

SilkePilon commented:
Lint seems to be failing atm

SilkePilon committed:
refactor: remove unused imports in AppearanceSettings.svelte

Remove several unused imports from AppearanceSettings.svelte to clean
up the codebase. Specifically remove:
- platformName from platform.ts
- Tauri import and its related context variable
- slide transition from svelte/transition
- getContext from @gitbutler/shared/context

SilkePilon committed:
Merge pull request #14 from SilkePilon/refactor-remove-unused-imports-appearancesettings

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
Merge branch 'master' into master

PavelLaptev committed:
some style tweaks

- gb icon change
- title bar height made smaller

PavelLaptev committed:
Refactor WindowsTitleBar badge and update styles

- update color for the Badge component
- move title bar into a CSS variable
- fix paddings
- remove the app name

PavelLaptev committed:
ui fixes

- update controls. Made them less contrast
- remove dark mode class. no needed because variables should do the trick

PavelLaptev commented on apps/desktop/src/components/WindowsTitleBar.svelte:
```diff
@@ -556,7 +555,7 @@
 	.native-control-button:hover {
 		background-color: rgba(255, 255, 255, 0.1);
-	}
+	} 
```
might be auto-formatting. I'll clean this up

PavelLaptev committed:
remove social net icons

don't like how they look there - too big and outline icons would be nicer.

PavelLaptev commented:
@SilkePilon When you click on the menu item it opens the menu and when click again it should close the menu, not trigger it. Should work like a toggle:

https://github.com/user-attachments/assets/e9320949-00a2-4b87-a322-4df5038640ef

SilkePilon commented on packages/ui/src/lib/Badge.svelte:
```diff
@@ -32,7 +32,21 @@
 	<!-- svelte-ignore a11y_no_static_element_interactions -->
 	<!-- svelte-ignore a11y_click_events_have_key_events -->
-	<div class="badge {style} {kind} {size}-size" class:reversedDirection {onclick}>
+	<div
+		class="badge"
+		class:neutral={style === 'neutral'}
+		class:pop={style === 'pop'}
+		class:success={style === 'success'}
+		class:error={style === 'error'}
+		class:warning={style === 'warning'}
+		class:purple={style === 'purple'}
+		class:solid={kind === 'solid'}
+		class:soft={kind === 'soft'}
+		class:icon-size={size === 'icon'}
+		class:tag-size={size === 'tag'}
```
reverted

SilkePilon committed:
fix: simplify Badge component class logic

Refactor Badge.svelte to use string interpolation for class names
instead of multiple class:directive conditionals. This approach 
reduces code duplication and improves readability while maintaining
the same functionality.

SilkePilon committed:
feat: add dropdown visibility tracking

Implement state tracking for menu dropdowns in WindowsTitleBar. Add 
toggle functions for file, view, project, and help dropdowns to handle 
open/close states. Update menu item click handlers to close dropdowns 
after actions are performed, providing better UX by preventing menus 
from staying open after interaction.

SilkePilon commented:
> Another thing is that it looks like the custom title bar can maximize the window, but not minimize it:

Weird, this seems to work for me just fine.

SilkePilon committed:
Merge branch 'master' into master

SilkePilon committed:
refactor: streamline dropdown management in WindowsTitleBar

Refactor dropdown handling in WindowsTitleBar component to use a 
centralized data structure and generic toggle/close functions. This 
replaces multiple individual state variables and duplicate toggle 
functions with a single unified approach, making the code more
maintainable and reducing repetition while preserving the same
functionality.

SilkePilon committed:
Merge pull request #15 from SilkePilon/sp-branch-2

SilkePilon commented:
> @SilkePilon When you click on the menu item it opens the menu and when click again it should close the menu, not trigger it. Should work like a toggle:
> 
>  Screen.Recording.2025-06-25.072019.mp4



> @SilkePilon When you click on the menu item it opens the menu and when click again it should close the menu, not trigger it. Should work like a toggle:
> 
>  Screen.Recording.2025-06-25.072019.mp4

Change it so its an toggle

PavelLaptev committed:
remove extra platform check

we already doing this inside the component

PavelLaptev commented:
This issue is caused because you modified the dropdown component using global styles:

> Thanks! But it should be smarter, I think, because currently I have to click twice to trigger the dropdown (check the video below).

However, this is not actually a dropdown component. A proper dropdown component triggers the menu when clicking on the arrow and hide the menu just fine. But in your implementation not the arrow triggers the menu to show, it's the whole button. 

https://github.com/user-attachments/assets/ea4cdef0-6ed8-4afd-8974-efa6e999040d

Let's remove the dropdown component from here. It doesn’t make sense to use a component whose logic and styles have been overridden. Instead, it should be a custom element specifically designed for this custom title bar.

@mtsgrd you also asked if we can modify the dropdown button. In my view, it's not a dropdown button component anymore — it looks and work differently.
